File: PlatformSpellChecker/src/desktopMain/kotlin/com/darkrockstudios/libs/platformspellchecker/macos/NSSpellCheckerJNI.kt
package com.darkrockstudios.libs.platformspellchecker.macos

import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference

/**
 * JNA Interface for the native NSSpellChecker wrapper library.
 * This library uses out-parameters to avoid struct-by-value return issues.
 */
interface NSSpellCheckerJNI : Library {
    companion object {
        val INSTANCE: NSSpellCheckerJNI by lazy {
            // JNA will automatically look in the correct platform-specific directory
            // For macOS ARM64: darwin-aarch64/libNSSpellCheckerJNI.dylib
            // For macOS x86-64: darwin-x86-64/libNSSpellCheckerJNI.dylib
            Native.load("NSSpellCheckerJNI", NSSpellCheckerJNI::class.java)
        }

        // NSNotFound constant
        const val NSNotFound = Long.MAX_VALUE
    }

    /**
     * Check spelling of a string starting at a given offset.
     * @param text The text to check (UTF-8 encoded)
     * @param startingOffset The offset to start checking from
     * @param outLocation Output: location of the first misspelled word (NSNotFound if none)
     * @param outLength Output: length of the misspelled word
     */
    fun checkSpellingOfString(
        text: String,
        startingOffset: Long,
        outLocation: LongByReference,
        outLength: LongByReference
    )

    /**
     * Check spelling with language and wrap options.
     * @param text The text to check (UTF-8 encoded)
     * @param startingOffset The offset to start checking from
     * @param language The language to use for spell checking (can be null)
     * @param wrapFlag Whether to wrap around to the beginning when reaching the end
     * @param outLocation Output: location of the first misspelled word
     * @param outLength Output: length of the misspelled word
     */
    fun checkSpellingOfStringWithOptions(
        text: String,
        startingOffset: Long,
        language: String?,
        wrapFlag: Int,
        outLocation: LongByReference,
        outLength: LongByReference
    )

    /**
     * Find the range of a misspelled word in a string.
     * @param text The text to check (UTF-8 encoded)
     * @param startingOffset The offset to start checking from
     * @param language The language to use (can be null)
     * @param wrapFlag Whether to wrap around
     * @param outLocation Output: location of the misspelled word
     * @param outLength Output: length of the misspelled word
     * @param outWordCount Output: word count (can be null)
     */
    fun rangeOfMisspelledWord(
        text: String,
        startingOffset: Long,
        language: String?,
        wrapFlag: Int,
        outLocation: LongByReference,
        outLength: LongByReference,
        outWordCount: LongByReference?
    )

    /**
     * Get spelling suggestions for a word.
     * @param word The word to get suggestions for (UTF-8 encoded)
     * @param language The language to use (can be null)
     * @return Comma-separated list of suggestions (caller must free with freeMemory)
     */
    fun getSuggestions(word: String, language: String?): Pointer?

    /**
     * Check if a word is in the dictionary (correctly spelled).
     * @param word The word to check (UTF-8 encoded)
     * @param language The language to use (can be null)
     * @return 1 if the word is correct, 0 if misspelled
     */
    fun isWordInDictionary(word: String, language: String?): Int

    /**
     * Learn a new word (add to user dictionary).
     * @param word The word to learn (UTF-8 encoded)
     */
    fun learnWord(word: String)

    /**
     * Unlearn a word (remove from user dictionary).
     * @param word The word to unlearn (UTF-8 encoded)
     */
    fun unlearnWord(word: String)

    /**
     * Ignore a word for the current document/session.
     * @param word The word to ignore (UTF-8 encoded)
     */
    fun ignoreWord(word: String)

    /**
     * Get available languages.
     * @return Comma-separated list of language codes (caller must free with freeMemory)
     */
    fun getAvailableLanguages(): Pointer?

    /**
     * Set the language for spell checking.
     * @param language The language code to set
     * @return 1 if successful, 0 if failed
     */
    fun setLanguage(language: String): Int

    /**
     * Get the current language.
     * @return The current language code (caller must free with freeMemory)
     */
    fun getCurrentLanguage(): Pointer?

    /**
     * Check grammar in a string.
     * @param text The text to check (UTF-8 encoded)
     * @param startingOffset The offset to start checking from
     * @param language The language to use (can be null)
     * @param outLocation Output: location of the grammar error
     * @param outLength Output: length of the grammar error
     */
    fun checkGrammar(
        text: String,
        startingOffset: Long,
        language: String?,
        outLocation: LongByReference,
        outLength: LongByReference
    )

    /**
     * Count continuous spell checking errors from a given offset.
     * @param text The text to check (UTF-8 encoded)
     * @param startingOffset The offset to start from
     * @param language The language to use (can be null)
     * @return The number of errors found
     */
    fun countContinuousSpellCheckingErrors(
        text: String,
        startingOffset: Long,
        language: String?
    ): Long

    /**
     * Free memory allocated by this library.
     * @param ptr The pointer to free
     */
    fun freeMemory(ptr: Pointer)
}

/**
 * Helper extension to convert Pointer to String and free memory
 */
fun Pointer?.toStringAndFree(): String {
    if (this == null) return ""
    return try {
        getString(0, "UTF-8")
    } finally {
        NSSpellCheckerJNI.INSTANCE.freeMemory(this)
    }
}

File: PlatformSpellChecker/src/desktopMain/native/NSSpellCheckerJNI.m
/**
 * NSSpellCheckerJNI.m
 *
 * Native Objective-C wrapper for NSSpellChecker that avoids struct-by-value returns
 * by using out-parameters instead. This solves JNA's limitations with NSRange returns.
 */

#import <Foundation/Foundation.h>
#import <AppKit/AppKit.h>

// Initialize the shared NSSpellChecker instance
static NSSpellChecker* getSharedSpellChecker() {
    static NSSpellChecker* sharedChecker = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedChecker = [NSSpellChecker sharedSpellChecker];
    });
    return sharedChecker;
}

// Check spelling of a string starting at a given offset
// Returns via out-parameters to avoid NSRange struct return issues with JNA
__attribute__((visibility("default")))
void checkSpellingOfString(const char* text,
                          long startingOffset,
                          long* outLocation,
                          long* outLength) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* string = [NSString stringWithUTF8String:text];

        NSRange range = [checker checkSpellingOfString:string
                                           startingAt:(NSInteger)startingOffset];

        *outLocation = (long)range.location;
        *outLength = (long)range.length;
    }
}

// Check spelling with language and wrap mode options
__attribute__((visibility("default")))
void checkSpellingOfStringWithOptions(const char* text,
                                     long startingOffset,
                                     const char* language,
                                     int wrapFlag,
                                     long* outLocation,
                                     long* outLength) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* string = [NSString stringWithUTF8String:text];
        NSString* lang = language ? [NSString stringWithUTF8String:language] : nil;

        NSRange range = [checker checkSpellingOfString:string
                                           startingAt:(NSInteger)startingOffset
                                             language:lang
                                                 wrap:(BOOL)wrapFlag
                                   inSpellDocumentWithTag:0
                                                wordCount:NULL];

        *outLocation = (long)range.location;
        *outLength = (long)range.length;
    }
}

// Find the range of a misspelled word in a string
__attribute__((visibility("default")))
void rangeOfMisspelledWord(const char* text,
                          long startingOffset,
                          const char* language,
                          int wrapFlag,
                          long* outLocation,
                          long* outLength,
                          long* outWordCount) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* string = [NSString stringWithUTF8String:text];

        // Use the simpler checkSpellingOfString:startingAt: method
        // Note: This doesn't support language specification directly
        NSRange range = [checker checkSpellingOfString:string
                                           startingAt:(NSInteger)startingOffset];

        *outLocation = (long)range.location;
        *outLength = (long)range.length;
        if (outWordCount) {
            // Count words in the range
            if (range.location != NSNotFound) {
                *outWordCount = 1; // Simple count for now
            } else {
                *outWordCount = 0;
            }
        }
    }
}

// Get spelling suggestions for a word
// Returns a comma-separated list of suggestions (caller must free)
__attribute__((visibility("default")))
char* getSuggestions(const char* word, const char* language) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* wordString = [NSString stringWithUTF8String:word];
        NSString* lang = language ? [NSString stringWithUTF8String:language] : nil;

        NSArray<NSString*>* suggestions;
        if (lang) {
            // Use newer API if language is specified
            suggestions = [checker guessesForWordRange:NSMakeRange(0, [wordString length])
                                              inString:wordString
                                              language:lang
                                    inSpellDocumentWithTag:0];
        } else {
            // Fall back to simpler API (deprecated but still functional)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
            suggestions = [checker guessesForWord:wordString];
#pragma clang diagnostic pop
        }

        if (!suggestions || [suggestions count] == 0) {
            return strdup("");
        }

        // Join suggestions with comma
        NSString* joined = [suggestions componentsJoinedByString:@","];
        return strdup([joined UTF8String]);
    }
}

// Check if a word is in the dictionary (correctly spelled)
__attribute__((visibility("default")))
int isWordInDictionary(const char* word, const char* language) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* wordString = [NSString stringWithUTF8String:word];
        NSString* lang = language ? [NSString stringWithUTF8String:language] : nil;

        // Use checkSpellingOfString to determine if word is correct
        NSRange range = [checker checkSpellingOfString:wordString
                                           startingAt:0
                                             language:lang
                                                 wrap:NO
                                   inSpellDocumentWithTag:0
                                                wordCount:NULL];

        // If NSNotFound is returned, the word is spelled correctly
        return (range.location == NSNotFound) ? 1 : 0;
    }
}

// Learn a new word (add to user dictionary)
__attribute__((visibility("default")))
void learnWord(const char* word) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* wordString = [NSString stringWithUTF8String:word];
        [checker learnWord:wordString];
    }
}

// Unlearn a word (remove from user dictionary)
__attribute__((visibility("default")))
void unlearnWord(const char* word) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* wordString = [NSString stringWithUTF8String:word];
        [checker unlearnWord:wordString];
    }
}

// Ignore a word for the current document/session
__attribute__((visibility("default")))
void ignoreWord(const char* word) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* wordString = [NSString stringWithUTF8String:word];
        [checker ignoreWord:wordString inSpellDocumentWithTag:0];
    }
}

// Get available languages as a comma-separated list (caller must free)
__attribute__((visibility("default")))
char* getAvailableLanguages() {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSArray<NSString*>* languages = [checker availableLanguages];

        if (!languages || [languages count] == 0) {
            return strdup("");
        }

        NSString* joined = [languages componentsJoinedByString:@","];
        return strdup([joined UTF8String]);
    }
}

// Set the language for spell checking
__attribute__((visibility("default")))
int setLanguage(const char* language) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* lang = [NSString stringWithUTF8String:language];

        BOOL result = [checker setLanguage:lang];
        return result ? 1 : 0;
    }
}

// Get the current language
__attribute__((visibility("default")))
char* getCurrentLanguage() {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* language = [checker language];

        if (!language) {
            return strdup("");
        }

        return strdup([language UTF8String]);
    }
}

// Check grammar in a string
__attribute__((visibility("default")))
void checkGrammar(const char* text,
                 long startingOffset,
                 const char* language,
                 long* outLocation,
                 long* outLength) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* string = [NSString stringWithUTF8String:text];
        NSString* lang = language ? [NSString stringWithUTF8String:language] : nil;

        NSArray* details = nil;
        NSRange range = [checker checkGrammarOfString:string
                                          startingAt:(NSInteger)startingOffset
                                            language:lang
                                                wrap:NO
                                  inSpellDocumentWithTag:0
                                             details:&details];

        *outLocation = (long)range.location;
        *outLength = (long)range.length;
    }
}

// Count continuous spell checking errors from a given offset
__attribute__((visibility("default")))
long countContinuousSpellCheckingErrors(const char* text,
                                       long startingOffset,
                                       const char* language) {
    @autoreleasepool {
        NSSpellChecker* checker = getSharedSpellChecker();
        NSString* string = [NSString stringWithUTF8String:text];
        NSString* lang = language ? [NSString stringWithUTF8String:language] : nil;

        NSInteger count = [checker countWordsInString:string
                                             language:lang];

        return (long)count;
    }
}

// Free allocated memory (helper for JNA)
__attribute__((visibility("default")))
void freeMemory(void* ptr) {
    if (ptr) {
        free(ptr);
    }
}